webapp/src/auth.ts

```diff
@@ -299,15 +330,70 @@ export class Component<TProps, TState> extends data.Component<TProps, TState> {
     }
 }
 
+export async function updateUserPreferencesAsync(newPref: Partial<UserPreferences>) {
+    // Update our local state
+    internalPrefUpdateAndInvalidate(newPref)
+
+    // If we're not logged in, non-persistent local state is all we'll use
+    if (!await loggedIn()) { return; }
+
+    // If the user is logged in, save to cloud
+    const result = await apiAsync<UserPreferences>('/api/user/preferences', newPref);
+    if (result.success) {
+        pxt.debug("Updating local user preferences w/ cloud data after result of POST")
+        // Set user profile from returned value so we stay in-sync
+        internalPrefUpdateAndInvalidate(result.resp)
+    } else {
+        pxt.reportError("identity", "update preferences failed failed", result as any);
+    }
+}
+
+function internalPrefUpdateAndInvalidate(newPref: Partial<UserPreferences>) {
+    // TODO is there a generic way to do this so we don't need to add new branches
+    //  for each field that changes?
+
+    // remember old
+    const oldPref = state_.preferences ?? DEFAULT_USER_PREFERENCES()
+    // update
+    state_.preferences = { ...oldPref, ...newPref }
+    // invalidate fields that change
+    if (oldPref?.highContrast !== state_.preferences?.highContrast) {
+        data.invalidate(HIGHCONTRAST)
+    }
+    if (oldPref?.language !== state_.preferences?.language) {
+        data.invalidate(LANGUAGE)
+    }
+}
+
+let initialUserPreferences_: Promise<UserPreferences | undefined> = undefined;
+export async function initialUserPreferences(): Promise<UserPreferences | undefined> {
```

Comment:
@eanders-MS , this is the pattern I'm using for getting preferences. We first wait for a loggin check and then we do a single fetch of the preferences and subsequent calls use this. I think this makes sense.

Eventually we should return more data from the serve in the initial payload but I think this works well for now.